feat: Support custom rulesets (#6083)
PHP-CS-Fixer/PHP-CS-Fixer: doc/config.rst

        ->setIndent("\t")
        ->setLineEnding("\r\n")
    ;

> Using custom rule sets. In this example `\MyNameSpace\MyRuleSetClass` must implement `\PhpCsFixer\RuleSet\RuleSetDescriptionInterface`.

Reviewer: ```suggestion
It's possible to register custom rule sets, which makes it easier to reuse custom configuration between multiple projects. If you have prepared rule set, you can register it and then enable it in the rules. Custom rule sets (in this example `\MyNameSpace\MyRuleSetClass`) must implement `\PhpCsFixer\RuleSet\RuleSetDescriptionInterface`.
```

PS. I am not sure if this should be added here, or in `doc/custom_rules.rst` 🤔.

.. code-block:: php

    <?php

    $config = new PhpCsFixer\Config();
    return $config
        ->registerCustomRuleSets([
            '@MyRuleSet' => \MyNameSpace\MyRuleSetClass::class,
        ])
        ->setRules([
            '@MyRuleSet' => true,
        ])

>     ;

Reviewer: ```suggestion
    ;

ℹ️ If you use other `\PhpCsFixer\ConfigInterface` implementation than built-in one, you can register custom rule sets using `\PhpCsFixer\RuleSet\RuleSets::registerRuleSet()` just before the configuration. Currently `\PhpCsFixer\Config:: registerCustomRuleSets()` is not part of the `ConfigInterface` because it would be a BC-break - it will be added there in next major version of the Fixer.
```

Author: Done

        ->setIndent("\t")
        ->setLineEnding("\r\n")
    ;

> It's possible to register custom rule sets, which makes it easier to reuse custom configuration between multiple projects. If you have prepared rule set, you can register it and then enable it in the rules. Custom rule sets (in this example `\MyNameSpace\MyRuleSetClass`) must implement `\PhpCsFixer\RuleSet\RuleSetDescriptionInterface`.

Reviewer 2: First of all, big thanks @niklam to take effort on this PR, and thanks @Wirone to take it back from dead and collaborating with @niklam to move this PR further 🍻 !

Yet here, before next code changes, I would love to align on the approach to how we would like to achieve the custom rulesets first.

.. code-block:: php

    <?php

    return (new PhpCsFixer\Config())
        ->registerCustomRuleSets([
            '@MyRuleSet' => MyNameSpace\MyRuleSetClass::class,
        ])
        ->setRules([
            '@MyRuleSet' => true,
        ])
    ;

> ℹ️ If you use other `\PhpCsFixer\ConfigInterface` implementation than built-in one, you can register custom rule sets using `\PhpCsFixer\RuleSet\RuleSets::registerRuleSet()` just before the configuration. Currently `\PhpCsFixer\Config::registerCustomRuleSets()` is not part of the `ConfigInterface` because it would be a BC-break - it will be added there in the next major version of the Fixer.

Reviewer 2: do not promote calling `RuleSets::registerRuleSet` directly - it's internal class

Reviewer: I believe we should have information how to use it when custom config implementation is used (and there's no API for registering ruleset). Maybe we just should explicitly state it's a part of internal API and does not have BC-promise?

Reviewer 2: or maybe if one is using _custom ruleset_ with _custom config_, it's OK to expect them to implement future-ready Config method

Reviewer: But it does not solve anything since it's not part of the contract. `Config` just works as a proxy to the static call, but in the future it won't require any changes if people stick with built-in implementation (even if it changes internal implementation to store custom rulesets to be read later in the process, rather than doing static call to `RuleSets`). Adding it in custom `ConfigInterface` implementation would only move static call from config file to that implementation - is it worth it?

Reviewer: It was reworked and now it's based on dedicated contract (`CustomRulesetsAwareConfigInterface`) and static calls are not needed anymore, only config class that implements the contract.

        ->setIndent("\t")
        ->setLineEnding("\r\n")
    ;

It's possible to register custom rule sets, which makes it easier to reuse custom configuration between multiple projects. If you have prepared rule set, you can register it and then enable it in the rules. Custom rule sets (in this example `\MyNameSpace\MyRuleSetClass`) must implement `\PhpCsFixer\RuleSet\RuleSetDescriptionInterface`.

.. code-block:: php

    <?php

>     $config = new PhpCsFixer\Config();
>     return $config

Reviewer 3: ```suggestion
    return (new PhpCsFixer\Config())
```

        ->setIndent("\t")
        ->setLineEnding("\r\n")
    ;

> It's possible to register custom rule sets, which makes it easier to reuse custom configuration between multiple projects. If you have prepared rule set, you can register it, and then enable it in the rules. Custom rule sets (in this example ``\MyNameSpace\MyRuleSetClass``) must implement ``\PhpCsFixer\RuleSet\RuleSetDescriptionInterface``.

Reviewer 2: I will externalize

.. code-block:: php

    <?php

    return (new PhpCsFixer\Config())

>         ->registerCustomRuleSets([

Reviewer 2: I hear you.

I still see the value in eating our own dog food, as tests didn't detected everything we detected with review, and maybe testing it out in action should show us more things.

ok to move on for now

            MyNameSpace\MyRuleSetClass::class, // It identifies itself as '@MyRuleSet'
        ])
        ->setRules([
            '@MyRuleSet' => true,
        ])
    ;

> ℹ️ If you use other ``\PhpCsFixer\ConfigInterface`` implementation than built-in one, you can register custom rule sets using ``\PhpCsFixer\RuleSet\RuleSets::registerRuleSet()`` just before the configuration. Currently ``\PhpCsFixer\Config::registerCustomRuleSets()`` is not part of the ``ConfigInterface`` because it would be a BC-break - it will be added there in the next major version of the Fixer.

Reviewer: Alternative approach is to introduce additional interface and implement it in default `Config` class, just like it's done in #7777 with `ParallelRunnerConfigInterface` 🙂. This way we would have out-of-the-box support for it in our config implementation, and anyone using custom config classes could implement this interface too. Then `RuleSets::registerRuleSet()` wouldn't be needed anymore.

Reviewer: It was done with `CustomRulesetsAwareConfigInterface`.
